hw09_struct_validator/validator/in_rule.go

```diff
@@ -0,0 +1,78 @@
+package validator
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+)
+
+// InRule проверяет входит ли значение в указанное множество.
+type InRule struct {
+	values []interface{}
+}
+
+func (m *InRule) Init(kind reflect.Kind, args []string) error {
+	m.values = make([]interface{}, len(args))
+	if !m.supports(kind) {
+		return ErrSupportArgType
+	}
+	if len(args) == 0 {
+		return ErrWrongArgsList
+	}
+	for i, arg := range args {
+		var (
+			v   interface{}
+			err error
+		)
+		if kind >= reflect.Int && kind <= reflect.Int64 {
+			v, err = strconv.ParseInt(arg, 10, 64)
+		}
+		if kind >= reflect.Uint && kind <= reflect.Uint64 {
+			v, err = strconv.ParseInt(arg, 10, 64)
+		}
+		if kind >= reflect.Float32 && kind <= reflect.Float64 {
+			v, err = strconv.ParseFloat(args[0], 64)
+		}
+		if kind == reflect.String {
+			v = arg
+		}
+		if err != nil {
+			return err
+		}
+		m.values[i] = v
+	}
+	return nil
+}
+
+func (m InRule) Check(val reflect.Value) error {
+	kind := val.Kind()
+	if !m.supports(kind) {
+		return ErrSupportArgType
+	}
+	for _, iv := range m.values {
```

Comment:
Критично:

А что мы проверяем в этом цикле? Если честно - непонятно. Такое чувство что нам достаточно чтобы первый элемент массива имел нужный тип. но это неверно. Нам нужно чтобы все элементы массива имели нужный тип. Приэтом, либо числа, либо строки.

Reply:
Тут мысль такая. По условию ДЗ, если поле структуры является массивом или слайсом, то при валидации к каждому элементу массива необходимо применить правило, которое указано в тегах и собрать оттуда ошибки. Это условие одинаково для всех типов правил и всех типов полей (в т.ч. типа структур или встроенных структур). Поэтому я эту логику решил вынести из чекеров правил и реализовать в процедуре валидации. 

```
hw09_struct_validator/validator/validator.go

	switch fVal.Kind() { //nolint:exhaustive // есть default
		case reflect.Slice, reflect.Array:
			rules, _ := ruleSet.(Rules)
			for i := 0; i < fVal.Len(); i++ {
				errorSet = append(errorSet, checkValue(fVal.Index(i), rules, append(names, strconv.Itoa(i))...)...)
			}

```

Т.е. в метод Check может попасть только значение скалярнго типа и я просто ищу его в массиве из Tag. 
Привидение к нужному типу значений массива m.values (это из тегов) идет в методе init.

Сравниваются значения, а не типы.

Для демонстрации того, что это работает добавил эти кейсы в тесты для строкового и целочисленного типов (UserRoles, Scores)

Reply:
Окей, но зачем на пробегать по всему массиву?

Логика у нас тут получается такая:

1. Берём первый элемент.
2. Проверяем что он тот самый. Если это так, завершаемся. Иначе берём следующий элемент и проверяем его. 
3. Если не нашли то ошибка.

Так?

Теперь, рассмотри возможные варианты проверяемого массива:
1. []int {1, 2, 3}
2. []bool{true, false}
3. []interface{1, "2", "foo"}

Что будет в каждом из них?

Reply:
```suggestion
	Окей, но зачем на пробегать по всему массиву?
```
Ну по заданию было собирать все ошибки, я и решил, что надо и все ошибки собрать в т.ч. и в массиве

```suggestion
	[]bool{true, false}
```
Задание ограничено было числовыми и строковыми типами

```suggestion
	[]interface{1, "2", "foo"}
```
Ну это последний удар, тут все рушится и надо хранить возможные значения в строках, а приведение к нужному типу в Check